tests/e2e/etcdmigration/etcdmigration_test.go

```diff
@@ -0,0 +1,214 @@
+package etcdmigration
+
+import (
+	"flag"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/k3s-io/k3s/tests"
+	"github.com/k3s-io/k3s/tests/e2e"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// Valid nodeOS: bento/ubuntu-24.04, opensuse/Leap-15.6.x86_64
+var nodeOS = flag.String("nodeOS", "bento/ubuntu-24.04", "VM operating system")
+var ci = flag.Bool("ci", false, "running on CI")
+var local = flag.Bool("local", false, "deploy a locally built K3s binary")
+
+// Environment Variables Info:
+// E2E_RELEASE_VERSION=v1.23.1+k3s2 or nil for latest commit from master
+
+// This test suite is used to verify that K3s can migrate from SQLite to etcd datastore
+// without losing data or functionality.
+
+func Test_E2EEtcdMigration(t *testing.T) {
+	RegisterFailHandler(Fail)
+	flag.Parse()
+	suiteConfig, reporterConfig := GinkgoConfiguration()
+	suiteConfig.Timeout = 30 * time.Minute
+	RunSpecs(t, "Etcd Migration Test Suite", suiteConfig, reporterConfig)
+}
+
+var tc *e2e.TestConfig
+
+// StartK3sWithSQLite starts K3s with SQLite as the datastore
+func StartK3sWithSQLite(nodes []e2e.VagrantNode) error {
+	for _, node := range nodes {
+		var startCmd string
+		if strings.Contains(node.String(), "server") {
+			startCmd = "systemctl start k3s"
+		} else {
+			startCmd = "systemctl start k3s-agent"
+		}
+		if _, err := node.RunCmdOnNode(startCmd); err != nil {
+			return &e2e.NodeError{Node: node, Cmd: startCmd, Err: err}
+		}
+	}
+	return nil
+}
+
+// MigrateToEtcd migrates the K3s datastore from SQLite to etcd
+func MigrateToEtcd(server e2e.VagrantNode) error {
+	// Stop K3s service
+	stopCmd := "systemctl stop k3s"
+	if _, err := server.RunCmdOnNode(stopCmd); err != nil {
+		return &e2e.NodeError{Node: server, Cmd: stopCmd, Err: err}
+	}
+
+	// Update config to use etcd
+	configCmd := "echo 'cluster-init: true' >> /etc/rancher/k3s/config.yaml"
+	if _, err := server.RunCmdOnNode(configCmd); err != nil {
+		return &e2e.NodeError{Node: server, Cmd: configCmd, Err: err}
+	}
+
+	// Start K3s with etcd
+	startCmd := "systemctl start k3s"
+	if _, err := server.RunCmdOnNode(startCmd); err != nil {
+		return &e2e.NodeError{Node: server, Cmd: startCmd, Err: err}
+	}
+
+	return nil
+}
+
+// KillK3sCluster kills the K3s cluster
+func KillK3sCluster(nodes []e2e.VagrantNode) error {
```

Comment:
We should move this function inside the e2e testutils.go file and call that for both startup and this new test function. Probably [before this function](https://github.com/k3s-io/k3s/blob/master/tests/e2e/testutils.go#L304).

Reply:
done